File: backend/src/main/java/br/ufpr/sept/so2/modules/solicitacoes/application/WorkflowJsonParser.java
package br.ufpr.sept.so2.modules.solicitacoes.application;

import br.ufpr.sept.so2.modules.solicitacoes.domain.WorkflowDefinicao;
import br.ufpr.sept.so2.shared.domain.exception.DadoInvalidoException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;

import java.util.LinkedHashMap;
import java.util.Map;

@Component
public class WorkflowJsonParser {

    private final ObjectMapper objectMapper;

    public WorkflowJsonParser(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    public WorkflowDefinicao parse(String workflowJson) {
        try {
            JsonNode root = objectMapper.readTree(workflowJson);
            String inicial = root.path("initial").asText(null);
            Map<String, Map<String, String>> transicoes = new LinkedHashMap<>();
            root.path("states").fields().forEachRemaining(estado -> {
                Map<String, String> acoes = new LinkedHashMap<>();
                estado.getValue().path("on").fields()
                        .forEachRemaining(acao -> acoes.put(acao.getKey(), acao.getValue().asText()));
                transicoes.put(estado.getKey(), acoes);
            });
            return new WorkflowDefinicao(inicial, transicoes);
        } catch (DadoInvalidoException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new DadoInvalidoException("workflow_json inválido.");
        }
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/solicitacoes/application/WorkflowJsonParser.kt
package br.ufpr.sept.so2.modules.solicitacoes.application

import br.ufpr.sept.so2.modules.solicitacoes.domain.WorkflowDefinicao
import br.ufpr.sept.so2.shared.domain.exception.DadoInvalidoException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Component
import java.util.LinkedHashMap

@Component
class WorkflowJsonParser(
    private val objectMapper: ObjectMapper,
) {

    fun parse(workflowJson: String?): WorkflowDefinicao {
        try {
            val root = objectMapper.readTree(workflowJson)
            val inicial = root.path("initial").asText(null)
            val transicoes = LinkedHashMap<String, Map<String, String>>()
            root.path("states").fields().forEachRemaining { estado ->
                val acoes = LinkedHashMap<String, String>()
                estado.value.path("on").fields()
                    .forEachRemaining { acao -> acoes[acao.key] = acao.value.asText() }
                transicoes[estado.key] = acoes
            }
            return WorkflowDefinicao(inicial, transicoes)
        } catch (ex: DadoInvalidoException) {
            throw ex
        } catch (_: Exception) {
            throw DadoInvalidoException("workflow_json inválido.")
        }
    }
}
